File: languages/golang/spammer.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"time"

	"github.com/DataDog/datadog-go/statsd"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
)

func main() {
	fmt.Printf("Waiting 10 seconds for agent to be ready\n")
	time.Sleep(10 * time.Second)
	fmt.Printf("Starting spammer at: %v\n", time.Now().Unix())
	tracer.Start()
	defer tracer.Stop()
	globalTags := []string{
		"language:golang",
		fmt.Sprintf("transport:%s", os.Getenv("TRANSPORT")),
		fmt.Sprintf("conc:%s", os.Getenv("CONCURRENT_SPAMMERS")),
		fmt.Sprintf("trunid:%s", os.Getenv("TRANSPORT_RUN_ID")),
		fmt.Sprintf("env:%s", os.Getenv("DD_ENV")),
		fmt.Sprintf("service:%s", os.Getenv("DD_SERVICE")),
		fmt.Sprintf("version:%s", os.Getenv("DD_VERSION")),
	}
	statsdClient, err := statsd.New("observer:8125", statsd.WithTags(globalTags))
	if err != nil {
		log.Fatal(err)
	}
	stressTest := &stressTest{statsdClient}

	statsdClient.Incr("transport_sample.run", nil, 1)
	spansCreated := int64(0)
	spanDiff := int64(0)
	previousSpanSubmit := int64(0)

	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, os.Interrupt)
	for {
		select {
		case <-sigs:

			// Submit last batch of live counting
			spanDiff := spansCreated - previousSpanSubmit
			statsdClient.Count("transport_sample.span_created", spanDiff, nil, 1)

			fmt.Printf("Finishing at: %v\n", time.Now().Unix())
			fmt.Printf("Spans created: %f\n", spansCreated)
	        statsdClient.Count("transport_sample.span_logged", spansCreated, nil, 1)
	        statsdClient.Incr("transport_sample.end", nil, 1)
			time.Sleep(5 * time.Second)
			statsdClient.Close()
			os.Exit(0)
		default:
			stressTest.createTrace()
			spansCreated = spansCreated + 2
			spanDiff = spansCreated - previousSpanSubmit

			if spanDiff > 499 {
				statsdClient.Count("transport_sample.span_created", spanDiff, nil, 1)
				previousSpanSubmit = spansCreated
			}
		}
	}
}

type stressTest struct {
	statsdClient *statsd.Client
}

// createTrace creates a trace with two spans,
// where "nested-spam" is a child of "spam".
// Each span creation sends an event to DogStatsD to
// help with throughput analysis.
func (st *stressTest) createTrace() {
	span := tracer.StartSpan("spam", tracer.ResourceName("spammer"))
	defer span.Finish()

	childSpan := tracer.StartSpan("nested-spam", tracer.ChildOf(span.Context()))
	time.Sleep(time.Millisecond)
	childSpan.Finish()
}

File: languages/java/src/main/java/com/datadog/transporttest/Spammer.java
package com.datadog.transporttest;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import com.timgroup.statsd.NonBlockingStatsDClientBuilder;
import com.timgroup.statsd.StatsDClient;
import datadog.opentracing.DDTracer;
import io.opentracing.Scope;
import io.opentracing.Span;
import io.opentracing.util.GlobalTracer;

public class Spammer {
    private static Logger log = LoggerFactory.getLogger(Spammer.class);

    public static void main(String[] args) throws InterruptedException {
        log.info("Sleeping for 10 seconds to wait for agent");
        Thread.sleep(10000);
        log.info("Starting spammer");

        String[] constantTags = new String[] {
                "language:java",
                "transport:" + System.getenv("TRANSPORT"),
                "conc:" + System.getenv("CONCURRENT_SPAMMERS"),
                "trunid:" + System.getenv("TRANSPORT_RUN_ID"),
                "env:" + System.getenv("DD_ENV"),
                "service:" + System.getenv("DD_SERVICE"),
                "version:" + System.getenv("DD_VERSION"),
                "tracer_version:" + datadog.trace.core.DDTraceCoreInfo.VERSION
        };

        StatsDClient observer = new NonBlockingStatsDClientBuilder()
                .hostname("observer")
                .port(8125)
                .constantTags(constantTags)
                // make sure all the metrics are sent before shutdown
                .blocking(true)
                // disable StatsD Client metrics
                .enableTelemetry(false)
                .enableAggregation(true)
                .errorHandler(e -> log.error("StatsDClient error: ", e))
                .build();

        final DDTracer tracer = DDTracer.builder()
                // disables tracer metrics
                .statsDClient(datadog.trace.api.StatsDClient.NO_OP)
                .build();
        GlobalTracer.registerIfAbsent(tracer);
        // register the same tracer with the Datadog API
        datadog.trace.api.GlobalTracer.registerIfAbsent(tracer);

        new Spammer(observer, tracer).run();
    }

    private final StatsDClient observer;
    private final DDTracer tracer;
    private long spansCreated;
    private long previousSpansCreated;

    private Spammer(StatsDClient observer, DDTracer tracer) {
        this.observer = observer;
        this.tracer = tracer;
    }

    private void run() throws InterruptedException {
        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override
            public void run() {
                incrementSpanCreated(1);
                observer.count("transport_sample.span_logged", spansCreated, 1.0);
                observer.increment("transport_sample.end");
                observer.close();
                tracer.close();
                log.info("Ended spammer");
            }
        });
        observer.increment("transport_sample.run");
        while (true) {
            final Span span = tracer.buildSpan("spam").withResourceName("spammer").start();
            try (final Scope scope = tracer.activateSpan(span)) {
                spansCreated += 1;
                final Span nestedSpan = tracer.buildSpan("nested-spam").asChildOf(span).start();
                try (final Scope nestedScope = tracer.activateSpan(nestedSpan)) {
                    spansCreated += 1;
                    Thread.sleep(1);
                }
                nestedSpan.finish();
            }
            span.finish();
            incrementSpanCreated(200);
        }
    }

    private void incrementSpanCreated(int limit) {
        long diff = spansCreated - previousSpansCreated;
        if (diff >= limit) {
            // Batch these so as to not overload dogstatsd
            observer.count("transport_sample.span_created", diff, 1.0);
            previousSpansCreated = spansCreated;
        }
    }
}
